### ISSUE TITLE
setFieldTouched, setFieldValue, etc. not working with arrays in Typescript anymore

### ISSUE THREAD

**Issue by mikecousins** [jaredpalmer/formik#1948]
### Current Behavior
![image](https://user-images.githubusercontent.com/725614/67737163-cb8be400-f9cf-11e9-9b39-f9afb5bd9cc8.png)

My initial values is a property called targets which contains an array of UITargetPositions.

![image](https://user-images.githubusercontent.com/725614/67737194-e2323b00-f9cf-11e9-8abb-8e2533920d67.png)

Calling setFieldValue or setFieldTouched results in a Typescript error that the string isn't one of the possible values.

### Expected behavior

No error should be thrown.

### Reproducible example

I can't get it to work with Typescript. For some reason when I changed it to .tsx it just gives me a white screen now, but here you go:
https://codesandbox.io/s/formik-codesandbox-template-iru3w

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Formik           | 2.0.3
| React            | 16.11.0
| TypeScript       | 3.6.4
| Browser          | all
| npm/Yarn         | yarn 1.19.1
| Operating System | Win 10 Pro

**Comment by sa9sha9** [comment]
i'm facing same issue :(

![スクリーンショット 2019-10-31 11 29 29](https://user-images.githubusercontent.com/14252323/67916850-753ec100-fbda-11e9-91f2-d754db05456b.png)

**Comment by fiendfiend** [comment]
it appears that setFieldValue is now attempting to be a bit too smart in terms of inferring the possible field values based on the generic form values parameter.

In the case of nested fields, this no longer works, since it doesn't attempt to traverse the possible object graph in order to derive all possible string permutations for fields - nor do I think it can or should achieve such a thing.

**Comment by fenech** [comment]
The issue is  https://github.com/jaredpalmer/formik/blob/master/src/types.tsx#L89
```typescript
setFieldValue(field: keyof Values & string, value: any, shouldValidate?: boolean): void;
```
`keyof Values` only includes top-level properties, so cannot be used for any nested structures.

The only way I can think of to make this work (without reverting to just `field: string`) would be to add a typed element further down the tree, which exposed a "scoped" version of `setFieldValue`:
```typescript
<Formik<Values>>
  ...
  <FieldValueSetter<NestedValues> name="foo.bar[0].baz">    
    {({ setFieldValue }: FieldValueSetterProps<NestedValues>) => ...
  </FieldValueSetter>
</Formik>
```
where `FieldValueSetterProps` included
```typescript
interface FieldValueSetterProps<NestedValues> {
  setFieldValue(field: keyof NestedValues & string, value: any, shouldValidate?: boolean): void;
}
```

**Comment by dvberzon** [comment]
Hit this problem as well. The new types effectively make:
`setFieldValue`
`setFieldError`
`setFieldTouched`

All incompatible with nested or array values.

As a workaround, for now, I'm using `as` with a string literal for one of the top level properties. Eg:

`setFieldValue('category.name' as 'category', myValue)`

However I don't think there is any harm in the type reverting back to `string`

**Comment by davidroeca** [comment]
I believe this is a duplicate of https://github.com/jaredpalmer/formik/issues/1698

**Comment by devhyunjae** [comment]
Any updates on this? Stuck in here!

**Comment by apieceofbart** [comment]
I totally agree this should be relaxed to be just string - it's nice to have the autocompletion for the simple case (non nested updates) but on the other hand types are just wrong currently

**Comment by CloudPower97** [comment]
Is this going to be fixed soon?

**Comment by use-strict** [comment]
Still an issue.

**Comment by FadiAboMsalam** [comment]
any fix for this 🤔  ?

**Comment by alealmarazdev** [comment]
This will be fixed?
